### ISSUE TITLE
Incorrect dynamic dispatch

### ISSUE THREAD

**Issue by magistere** [nim-lang/Nim#1373]
The following sample program causes problems with dynamic dispatching:
```nimrod
type
  PTree = ref object of PObject
  PEmpty = ref object of PTree
  PNode[T] = ref object of PTree
    val: T
    left: PTree
    right: PTree

method traverse(t: PTree) = discard # assert(false)
method traverse(t: PEmpty) = discard

method traverse(t: PNode) =
  traverse(t.left)
  echo(t.val)
  traverse(t.right)

let e = PEmpty()

when defined(v1):
  let tl = PNode[int](val: 5, left: e, right: e)
  let tr = PNode[int](val: 7, left: e, right: e)
  let t = PNode[float](val: 3, left: tl, right: tr)

  traverse(t)

echo "---"

when defined(v2):
  let tl1 = PNode[float](val: 5, left: e, right: e)
  let tr1 = PNode[int](val: 7, left: e, right: e)
  let t1 = PNode[int](val: 3, left: tl1, right: tr1)

  traverse(t1)
```
When `v1` is defined, method `traverse` for `PNode[int]` is not generated:
```
3.0
---
```
When `v2` is defined, method `traverse` for `PNode[float]` is not generated:
```
---
3
7
```
Only when both code fragments `v1` and `v2` are enabled, the program works as expected and produces **correct** result:
```
5
3.0
7
---
5.0
3
7
```

**Comment by Araq** [comment]
This behaviour has been fixed in 0.15.

**Comment by magistere** [comment]
Thanks for fixing this.
